enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/RemoteStore.java

```diff
@@ -123,20 +124,40 @@ private long getPullIndex( File storeDir ) throws IOException
                 lastTxId = tx.getCommitEntry().getTxId();
             }
 
-            if ( lastTxId < lastCleanTxId )
-            {
-                throw new IllegalStateException( "Metadata index was higher than transaction log index." );
-            }
-
-            // we don't want to pull a transaction we already have in the log, hence +1
-            return lastTxId + 1;
+            return Optional.of( lastTxId );
         }
     }
 
-    public CatchupResult tryCatchingUp( AdvertisedSocketAddress from, StoreId expectedStoreId, File storeDir ) throws StoreCopyFailedException, IOException
+    /**
+     * Later stages of the startup process require at least one transaction to
+     * figure out the mapping between the transaction log and the consensus log.
+     *
+     * If there are no transaction logs then we can pull from and including
+     * the index which the metadata store points to. This would be the case
+     * for example with a backup taken during an idle period of the system.
+     *
+     * However, if there are transaction logs then we want to find out where
+     * they end and pull from there, excluding the last one so that we do not
+     * get duplicate entries.
+     */
+    public CatchupResult tryCatchingUp( AdvertisedSocketAddress from, StoreId expectedStoreId ) throws StoreCopyFailedException, IOException
```

Comment:
I think we are still of by one. Assume that i = last index, T=tx, M=metadata.

In the scenario were tx logs are present we would before see:
    CLIENT -> Ti -> SERVER
    SERVER -> Ti+1 -> CLIENT

Now we have:
  CLIENT -> Ti-i -> SERVER
  SERVER -> Ti -> CLIENT.

In summary what we get from server now is one of:
Ti
Mi or Mi+1

What we want is:
Ti+1
Mi or Mi+1 

isn't it?

Reply:
Disregard this comment. I was wrong.